### ISSUE TITLE
We need to better test our new addons

### ISSUE THREAD

**Issue by lisa-wolfgang** [ScratchAddons/ScratchAddons#2622]
As our userbase increases, we need to reduce the number of release bugs to maintain reputability. https://github.com/ScratchAddons/ScratchAddons/issues/2614 should have been a pretty easy bug to catch before release, but it wasn't, and that's a little embarassing.

We need to start testing our addons more thoroughly before merging them or before release. A testing specification would help with this. Here are some items that could be included:
1. Test with every possible settings combination (if addon has settings)
2. Test with all other addons disabled
3. Test with all addons and addon settings enabled
4. Test dynamic disable/enable (the addon itself and all settings)
5. Test every possible scenario (toggle all functions in editor/webpage)

This could be added as a comment to the "Tests" section of the PR template, and the PR author should be required to indicate which tests were completed.

**Comment by Shluffy** [comment]
also, the pr template was not filled out, so they didn't mention anything about tests. possibly make sure that pr templates are being filled out and before the pr is merged have it tested again, as there are normally changes between the op and when it is merged

**Comment by apple502j** [comment]
Please do not joke. We'd rather fix bugs in dot release than having to test every possible scenarios.

**Comment by return-misaPuding** [comment]
I agree, but I think we should skip 2 and 4 because if there will be bugs
which we will find using one of these tests, we will probably find them
with 1 or 3 too

Dne čt 27. 5. 2021 13:57 uživatel lisa-wolfgang ***@***.***>
napsal:

**Comment by mxmou** [comment]
> Please do not joke. We'd rather fix bugs in dot release than having to test every possible scenarios.

This bug happens in a very common scenario (having *any* unread message), which makes it very easy to notice.

Of course there is no way to test how an addon behaves in every possible situation, but at least the most important ones should be tested. #2270 is an example of a bug that would be noticed immediately if just the minimum and maximum values for the zoom speed setting were tested.

**Comment by lisa-wolfgang** [comment]
As we were discussing in the Discord server, getting the beta-testing program going again a couple days before each release would be a good idea.

**Comment by apple502j** [comment]
This is not a bug.

**Comment by jeffalo** [comment]
This is something that may be possible to improve with an automated testing pipeline.

**Comment by apple502j** [comment]
@jeffalo which we do have - however, testing an extension isn't that easy (aside from basic lint results that people ignore before merging, and the unittest nobody runs)

**Comment by DNin01** [comment]
I've overseen several issues with my own projects after releasing updates when I think everything is okay and have realized how hard it is to catch everything. These oversights can be of things as simple as adding instead of subtracting.

Sometimes you forget to test something after a change, sometimes you don't even bother and just hope it will be fine, sometimes you don't even know that something can go wrong a certain way, and sometimes it's just impossible to test everything yourself.

This is why I think it is important to have a proper beta testing system, whether it be a team of volunteers or a beta channel that users can opt into. Crowds are the most powerful bug hunters. People discover bugs in games all the time, for example.

**Comment by DNin01** [comment]
Since this issue was created, each pull request must have two approvals before merging, which is progress to making sure things are tested. The Chrome Web Store also lets us gradually roll out updates so that if something isn't working, it has a smaller impact.

**Comment by DNin01** [comment]
*sigh* #7761

**Comment by WorldLanguages** [comment]
> *sigh* #7761

Differences between the development version and the releases will always be hard to catch.

**Comment by DNin01** [comment]
> > *sigh* #7761
> 
> 
> 
> Differences between the development version and the releases will always be hard to catch.

We _could_ theoretically have testers compile production builds and test them. But hopefully this won't happen again anyway…

**Comment by Samq64** [comment]
> > _sigh_ #7761
> 
> Differences between the development version and the releases will always be hard to catch.

Publishing beta releases even if they don't make it to the extension stores would have cought it if any firefox users tested it.
